scvi/inference/jvae_trainer.py

```diff
@@ -0,0 +1,386 @@
+                            latent_tensors = []
+                            for (i, (data, *_)) in enumerate(tensors):
+                                z = self.model.sample_from_posterior_z(
+                                    data, mode=i, deterministic=False
+                                )
+                                latent_tensors.append(z)
+
+                            # Train discriminator
+                            d_loss = self.loss_discriminator(
+                                [t.detach() for t in latent_tensors], True
+                            )
+                            d_loss *= self.kappa
+                            d_optimizer.zero_grad()
+                            d_loss.backward()
+                            d_optimizer.step()
+
+                            # Train generative model to fool discriminator
+                            fool_loss = self.loss_discriminator(latent_tensors, False)
+                            fool_loss *= self.kappa
+                            g_optimizer.zero_grad()
+                            fool_loss.backward()
+                            g_optimizer.step()
+
+                        # Train generative model
+                        g_loss = self.loss(tensors)
+                        g_optimizer.zero_grad()
+                        g_loss.backward()
+                        g_optimizer.step()
+
+                    if not self.on_epoch_end():
+                        break
+
+        self.model.eval()
+        self.training_time += (time.time() - begin) - self.compute_metrics_time
+        if self.verbose and self.frequency:
+            print(
+                "\nTraining time:  %i s. / %i epochs"
+                % (int(self.training_time), self.n_epochs)
+            )
+
+    def loss_discriminator(
+        self,
+        latent_tensors: List[torch.Tensor],
+        predict_true_class: bool = True,
+        return_details: bool = False,
+    ) -> Union[List[torch.Tensor], torch.Tensor]:
+        """
+        Compute the loss of the discriminator (either for the true labels or the fool labels)
+
+        :param latent_tensors: Tensors for each dataset of the latent space
+        :param predict_true_class: Specify if the loss aims at minimizing the accuracy or the mixing
+        :param return_details: Boolean used to inspect the loss values, return detailed loss for each dataset
+        :return: scalar loss if return_details is False, else list of scalar losses for each dataset
+        """
+        log_softmax = nn.LogSoftmax(dim=1)
+        n_classes = self.n_dataset
+        losses = []
+        for i, z in enumerate(latent_tensors):
+            cls_logits = log_softmax(self.discriminator(z))
+
+            if predict_true_class:
+                cls_target = torch.zeros(
+                    n_classes, dtype=torch.float32, device=z.device
+                )
+                cls_target[i] = 1.0
+            else:
+                cls_target = torch.ones(
```

Comment:
What is this for Predict_true_class ? To learn the optimal discriminator, you minimize the cross entropy between estimated and true class. To include in the adversarial loss (the H divergence) is it another term then ? I'm asking because it looks like this is where the error was last time we talked about it

Reply:
The error in the fish branch was that the adversarial loss was `-CE(true_class, prediction)`. With this if condition, you compute the optimal loss `+CE(true_class, prediction)` and the adversarial loss with `+CE(false_classes, prediction)`.
Is it right for you?